sdk/src/main/java/io/radar/sdk/util/RadarFileSystemLogBuffer.kt

```diff
@@ -0,0 +1,212 @@
+import io.radar.sdk.model.RadarLog
+import org.json.JSONObject
+import java.io.File
+import java.io.FileInputStream
+import java.io.FileOutputStream
+import java.util.*
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.locks.ReentrantLock
+
+/**
+ * Stores logs in files for persistent client-side logging
+ */
+internal class RadarFileSystemLogBuffer(
+    /**
+     * Application Context
+     */
+    context: Context,
+    /**
+     * Executor for writing files. This is used to ensure that log entries are written to the file in the order that
+     * they are received.
+     */
+    private val executor: ExecutorService = Executors.newSingleThreadExecutor()
+) : RadarLogBuffer {
+
+    private companion object {
+
+        /**
+         * Name of the directory in which the logs are stored
+         */
+        const val DIRECTORY = "RadarLogs"
+
+        /**
+         * Maximum number of logs in the buffer.
+         */
+        const val LOG_BUFFER_SIZE_LIMIT = 500
+    }
+
+    /**
+     * The directory where the log files are stored
+     */
+    private val directory = context.getDir(DIRECTORY, Context.MODE_PRIVATE)
+
+    /**
+     * Lock for handling changes to the files within [DIRECTORY]
+     */
+    private val directoryLock = ReentrantLock()
+
+    /**
+     * Contains the name of the current log file. Modifications to this value should occur within a [fileOp]
+     * operation.
+     */
+    private var currentFile: String
+
+    /**
+     * Number of log entries in the current log file. Modifications to this value should occur within a [fileOp]
+     * operation.
+     */
+    var size: Int
+        private set
+
+    init {
+        // First find current file
+        val files = directory.listFiles()
+        if (files?.isNotEmpty() == true) {
+            // This gets the latest file that was in use if it can, otherwise it returns the first one.
+            currentFile = files.maxByOrNull { getLastModified(it) }!!.name
+            // This is initialized to the number of lines in the current file
+            size = FileInputStream(File(directory, currentFile)).bufferedReader().readLines().size
+            if (files.size > 2) {
+                // This case likely means the app started and ended several times without completing a purge. To prevent
+                // excessive log files, purge them now.
+                purgeInactiveFiles()
+            }
+        } else {
+            // If no such file exists yet, create one.
+            currentFile = getNewFilePath()
+            size = 0
+        }
+    }
+
+    override fun write(level: Radar.RadarLogLevel, message: String) {
+        executor.submit {
+            if (isFull()) {
+                purgeInactiveFiles()
+            }
+            // Remove line-breaks to ensure each line in the file is a separate log event
+            val radarLog = RadarLog(level, message.replace('\n', '\t')).toJson()
```

Comment:
Are some messages multi-line?

Reply:
I kind of wonder if it even matters to do this?

Reply:
it matters for the way the log buffer works, which reads line-by-line. I don't think any messages are multi-line, but this protects against it.